tests/database-connection/mongodb-setup.js

```diff
@@ -0,0 +1,63 @@
+'use strict'
+
+const Fastify = require('fastify')
+const fp = require('fastify-plugin')
+
+const clean = require('mongo-clean')
+const { MongoClient } = require('mongodb')
+const { beforeEach, teardown } = require('tap')
+
+const App = require('../app')
+
+const url = 'mongodb://localhost/prod'
+const database = 'tests'
+
+let client
+
+beforeEach(async function () {
+  if (!client) {
+    client = await MongoClient.connect(url, {
+      w: 1,
+      useNewUrlParser: true
+    })
+  }
+  await clean(client.db(database))
+})
+
+teardown(async function () {
+  if (client) {
+    await client.close()
+    client = null
+  }
+})
+
+// Fill in this config with all the configurations
+// needed for testing the application
+function config () {
+  return {
+    mongodb: {
+      client,
+      database
+    }
+  }
+}
+
+// automatically build and tear down our instance
+function build (t) {
+  const app = Fastify()
+
+  // fastify-plugin ensures that all decorators
+  // are exposed for testing purposes, this is
+  // different from the production setup
+  app.register(fp(App), config())
+
+  // tear down our app after we are done
+  t.teardown(app.close.bind(app))
+
+  return app
+}
+
+module.exports = {
+  config,
+  build
+}
```

Comment:
This file mix tests flow (`beforeEach`, `teardown`) and utilities (`config`, `build`)
In this way importing this file has side effects that I would not suggest to the user

We should split the logic or create a factory function that accept the `t` `tap` object as input

Reply:
Can you elaborate? I took this example from Matteo message in the issue attached and looks pretty reasonable to me.

Reply:
In this file the `beforeEach` and `teardown` will always be the top-level `tap` test.

Let's say if someone `require` the file inside a `test`, it will mess up the test flow.
```js
const { test } = require('tap')

test('a', function() {
  const { build } = require('./mongodb-setup');
}) 

test('b', function() {
  const { build } = require('./mongodb-setup');
})
```
Simple Repro in [here](https://replit.com/@climba03003/tap-sideeffect)